operator/builtin/transformer/add/add_test.go

```diff
@@ -0,0 +1,204 @@
+		e.Record = map[string]interface{}{
+			"key": "val",
+			"nested": map[string]interface{}{
+				"nestedkey": "nestedval",
+			},
+		}
+		return e
+	}
+
+	cases := []testCase{
+		{
+			"add_value",
+			false,
+			newTestEntry,
+			func() *entry.Entry {
+				e := newTestEntry()
+				e.Record.(map[string]interface{})["new"] = "randomMessage"
+				return e
+			},
+		},
+		{
+			"add_expr",
+			false,
+			newTestEntry,
+			func() *entry.Entry {
+				e := newTestEntry()
+				e.Record.(map[string]interface{})["new"] = "val_suffix"
+				return e
+			},
+		},
+		{
+			"add_nest",
+			false,
+			newTestEntry,
+			func() *entry.Entry {
+				e := newTestEntry()
+				e.Record = map[string]interface{}{
+					"key": "val",
+					"nested": map[string]interface{}{
+						"nestedkey": "nestedval",
+					},
+					"new": map[interface{}]interface{}{
+						"nest": map[interface{}]interface{}{
+							"key": "val",
+						},
+					},
+				}
+				return e
+			},
+		},
+		{
+			"add_attribute",
+			false,
+			newTestEntry,
+			func() *entry.Entry {
+				e := newTestEntry()
+				e.Attributes = map[string]string{"new": "newVal"}
+				return e
+			},
+		},
+		{
+			"add_resource",
+			false,
+			newTestEntry,
+			func() *entry.Entry {
+				e := newTestEntry()
+				e.Resource = map[string]string{"new": "newVal"}
+				return e
+			},
+		},
+		{
+			"add_resource_expr",
+			false,
+			newTestEntry,
+			func() *entry.Entry {
+				e := newTestEntry()
+				e.Resource = map[string]string{"new": "val_suffix"}
+				return e
+			},
+		},
+	}
+	for _, tc := range cases {
+		t.Run("yaml/"+tc.name, func(t *testing.T) {
+			cfgFromYaml, _ := configFromFileViaYaml(path.Join(".", "testdata", fmt.Sprintf("%s.yaml", tc.name)))
+			cfgFromYaml.OutputIDs = []string{"fake"}
+			cfgFromYaml.OnError = "drop"
+			ops, err := cfgFromYaml.Build(testutil.NewBuildContext(t))
+			require.NoError(t, err)
+			op := ops[0]
+
+			add := op.(*AddOperator)
+			fake := testutil.NewFakeOutput(t)
+			add.SetOutputs([]operator.Operator{fake})
+
+			err = add.Process(context.Background(), tc.input())
+			if tc.expectErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				fake.ExpectEntry(t, tc.output())
+			}
+		})
+		t.Run("mapstructure/"+tc.name, func(t *testing.T) {
+			cfgFromMapstructure := defaultCfg()
+			configFromFileViaMapstructure(
+				path.Join(".", "testdata", fmt.Sprintf("%s.yaml", tc.name)),
+				cfgFromMapstructure,
+			)
+			cfgFromMapstructure.OutputIDs = []string{"fake"}
+			cfgFromMapstructure.OnError = "drop"
+			ops, err := cfgFromMapstructure.Build(testutil.NewBuildContext(t))
+			require.NoError(t, err)
+			op := ops[0]
+
+			remove := op.(*AddOperator)
+			fake := testutil.NewFakeOutput(t)
+			remove.SetOutputs([]operator.Operator{fake})
+			err = remove.Process(context.Background(), tc.input())
+			if tc.expectErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				fake.ExpectEntry(t, tc.output())
+			}
+		})
+	}
+}
```

Comment:
This is nice in that it is very "end-to-end", but I think it's clashing with the pattern we've already established in other operators. Let's pull this apart into two sets of tests:
1) Golden configs  (tests unmarshaling from yaml files into config structs)
2) Processing (configure an operator, then validate it processes an entry correctly)